miner/worker_builder.go

```diff
@@ -0,0 +1,472 @@
+package miner
+
+import (
+	"errors"
+	"math/big"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/txpool"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var (
+	errNonRevertingTxInBundleFailed = errors.New("non-reverting tx in bundle failed")
+	errBundlePriceTooLow            = errors.New("bundle price too low")
+)
+
+// commitWorkV2
+// TODO(roshan) consider whether to take bundle pool status as LOOP_WAIT condition
+func (w *worker) commitWorkV2(interruptCh chan int32, timestamp int64) {
+}
+
+// fillTransactions retrieves the pending bundles and transactions from the txpool and fills them
+// into the given sealing block. The selection and ordering strategy can be extended in the future.
+func (w *worker) fillTransactionsAndBundles(interruptCh chan int32, env *environment, stopTimer *time.Timer) error {
+	var (
+		pending   map[common.Address][]*txpool.LazyTransaction
+		localTxs  map[common.Address][]*txpool.LazyTransaction
+		remoteTxs map[common.Address][]*txpool.LazyTransaction
+		bundles   []*types.Bundle
+	)
+	{
+		// Split the pending transactions into locals and remotes
+		// Fill the block with all available pending transactions.
+		pending = w.eth.TxPool().Pending(false)
+
+		localTxs, remoteTxs = make(map[common.Address][]*txpool.LazyTransaction), pending
+		for _, account := range w.eth.TxPool().Locals() {
+			if txs := remoteTxs[account]; len(txs) > 0 {
+				delete(remoteTxs, account)
+				localTxs[account] = txs
+			}
+		}
+
+		bundles = w.eth.TxPool().PendingBundles(env.header.Number, env.header.Time)
+
+		log.Info("fill bundles and transactions", "bundles_count", len(bundles), "tx_count", len(pending))
+
+		// if no bundles, not necessary to fill transactions
+		if len(bundles) == 0 {
+			return errors.New("no bundles in bundle pool")
+		}
+	}
+
+	{
+		txs, bundle, err := w.generateOrderedBundles(env, bundles, pending)
+		if err != nil {
+			log.Error("fail to generate ordered bundles", "err", err)
+			return err
+		}
+
+		if err = w.commitBundles(env, txs, interruptCh, stopTimer); err != nil {
+			log.Error("fail to commit bundles", "err", err)
+			return err
+		}
+
+		env.profit.Add(env.profit, bundle.EthSentToSystem)
+	}
+
+	if len(localTxs) > 0 {
+		txs := newTransactionsByPriceAndNonce(env.signer, localTxs, env.header.BaseFee)
+		err := w.commitTransactions(env, txs, interruptCh, stopTimer)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(remoteTxs) > 0 {
+		txs := newTransactionsByPriceAndNonce(env.signer, remoteTxs, env.header.BaseFee)
+		err := w.commitTransactions(env, txs, interruptCh, stopTimer)
```

Comment:
To speed up the process of builder, I think a builder is running as a fast node, so please ensure not tier prefetch is enabled when doing `commitTransactions`.

Reply:
Done